src/discoverAutoScalingGroup.ts

```diff
@@ -0,0 +1,35 @@
+import * as sortBy from 'lodash/sortBy';
+import {AsgDiscoveryResponse, StateMachineInput} from './utils/handlerInputs';
+import {totalRunningExecutions} from "./aws/stepFunctions";
+import {getInstancesByTag} from './aws/ec2Instances';
+
+export async function handler(event: StateMachineInput): Promise<AsgDiscoveryResponse> {
+    const runningExecutionsPromise = totalRunningExecutions(event.stepFunctionArn)
+    const runningExecutions = await runningExecutionsPromise
+
+    if (runningExecutions !== 1) {
+        const error = `Expected to find one running execution (this one!) but there were ${runningExecutions}.`;
+        return { alreadyRunning: true };
+    }
+
+    if(event.asgName) {
+        console.log(`AutoScaling group ${event.asgName} specified as input. Moving on...`);
+        return { asgName: event.asgName, alreadyRunning: false };
+    }
+
+    const instances = await getInstancesByTag(event.autoScalingGroupDiscoveryTagKey);
+    
+    if(instances.length === 0) {
+        const error = `Could not find any instances to rotate with tag ${event.autoScalingGroupDiscoveryTagKey}.`;
+        throw new Error(error);
+    }
+
+    const oldestInstance = sortBy(instances, instance => instance.launchTime)[0];
+    console.log(`Found oldest instance ${oldestInstance.id} in AutoScaling group ${oldestInstance.autoScalingGroupName}`);
+    console.log(`Triggering rotation for AutoScaling group ${oldestInstance.autoScalingGroupName}`);
```

Comment:
It might make sense to have an age threshold, and not rotate any instance _at all_ if they are all younger than that threshold?

If the ENR Step Function is set to run quite frequently due to there being many Elasticsearch nodes in lots of clusters in our AWS account (perhaps running several times a day) - then if we were to later decide to shutdown most nodes (eg get rid of CODE, and scale PROD to be much smaller), the frequent runs of the ENR Step Function could end up rotating the few remaining nodes far too frequently, which would be unfortunate!

The age threshold could be something like 7 days?

Reply:
I totally agree, I've added the threshold configurable as a CFN parameter and defaulting to 7 days